Added Reactive Extensions Documentation

RLittlesII opened the pull request:
Added Extensions to the Main Nav

worldbeater commented:
It takes really a huge time to compile :smiley:

daveaglick commented:
> really a huge time to compile

How long we talking? I might use this site as a use case to do some profiling and optimization...

RLittlesII commented:
@daveaglick We are looking at about a solid hour for the +4000 files we added.  We are working to reduce the number of files.

https://dotnet.visualstudio.com/ReactiveUI/_build/results?buildId=5569&view=logs

glennawatson commented:
I will have a converter in a couple days to make it more usable.

RLittlesII commented:
@glennawatson I tried to use the converter.  I thought I pushed the results to the PR.  It still came out with a lot of extra folders, more than I would have expected.  Let me get to a windows box and I'll execute it again and verify and push the results.

RLittlesII commented:
@worldbeater what you see in the latest push is using the above DotNetConverter that @glennawatson wrote.  From here I am going to remove the majority of the testing links and try and clean up some of the duplicates based on generic typing and such.  I will also look into those odd characters you mentioned as I believe I see some in the output files.

daveaglick commented:
Are the new files source or Markdown (looks like MD from the commit, but just wanted to be sure). If it’s source, then that’s tricky and we’re partly at the mercy of Roslyn. If it’s Markdown, I’d be curious what the raw time just running those files through Markdig is in something like LINQPad - the difference is where we have potential to optimize.

RLittlesII commented:
It's just Markdown.  I will see if I can get that information for you.

RLittlesII commented:
@glennawatson @daveaglick The execution time of the program we are using to work with the MD files currently parses and dumps them in folders in approximately 25 sec.  The program we are using is up in https://github.com/reactiveui/website/pull/192#issuecomment-453769904

daveaglick commented:
From the Wyam Gitter:

Started looking at the build and I can already tell the API generation and Markdown parts are fine:
`Executed module Markdown in 2586 ms resulting in 2255 output document(s)`

It's still running but looks like it's getting held up in AutoLink which I guess makes some sense - that module reads each of the 2,272 input documents, forms a virtual DOM, queries the DOM, and then rewrites it as needed

Auto linking can be turned off with `DocsKeys.AutoLinkTypes` - I wonder if that would be an acceptable tradeoff /cc @glennawatson

Turning auto linking off lets me build the site much, much faster - though it still gets hung up in the Headings module for a while, which is similar

So it looks like there's some work to do around performance when using AngleSharp

daveaglick commented:
FWIW, I got the whole site from this PR to build with `DocsKeys.AutoLinkTypes` in about 8 minutes on my PC, so definitely doable if you're willing to forgo the automatic type linking. I'll continue to look into performance improvements in that module as well.

RLittlesII commented:
That is worth a lot.  What does forgoing the type linking lose us in terms of functionality?

daveaglick commented:
It won't automatically link the names of known types from the API in Markdown documents to their corresponding API page. I.e., with automatic linking turned on if there were a type named `Foo` in the API docs section, anywhere "Foo" appears in a Markdown document it would link to the `Foo` API page.

I'm hopefully I can figure out some performance enhancements for this, so maybe it's a way to get up and running with the extra documents from this PR with the expectation auto-linking can be turned back on later.

glennawatson commented:
@RLittlesII @daveaglick Go ahead and turn it off anyway locally while your fixing up the documentation.

daveaglick commented:
Just committed a performance improvement to `AutoLink` - I can now build this entire PR in about 6 minutes including automatic linking left on:

![2019-01-17_15h41_07](https://user-images.githubusercontent.com/1020407/51347641-b2342f80-1a6e-11e9-9efe-b32d3bc0b2ea.png)

(Yeah, that's not a great example since `Program` isn't actually an API type, but there _was_ a type in the API named `Program` and this linked to it so it's working).

I'll try to get a release with the fix out tomorrow or early next week.

RLittlesII commented:
@glennawatson @worldbeater I have removed all the odd characters that were identified.  I also removed the documents that seems like they were referencing tests.  Feel free to review.

glennawatson commented:
Lets wait until @daveaglick speed fixes are in before submitting this one.

worldbeater commented:
Really happy to see rx docs on ReactiveUI website, great job @RLittlesII 👍 
Compilation is really fast now! Although noticed a bug with page titles:
![image](https://user-images.githubusercontent.com/6759207/51402299-2dc0da00-1b5e-11e9-8a54-eed35c052728.png)

worldbeater commented on input/reactive-extensions/Add/CompositeDisposable.Add.md:
```diff
@@ -0,0 +1,67 @@
+
+Adds a disposable to the CompositeDisposable or disposes the disposable if the CompositeDisposable is disposed.
+
+**Namespace:**  [System.Reactive.Disposables](System.Reactive.Disposables\System.Reactive.Disposables.md)  
```
Namespace links seem to be broken, if I compile the website and click them, I get 404. Not sure to which location should those links point a user. Maybe worth making them not clickable?

glennawatson replied:
as mentioned by @daveaglick 
Set DocsKeys.ValidateRelativeLinks and it should run a link validator on all pipelines
That'll give you warnings, also set DocsKeys.ValidateLinksAsError to make invalid links generate errors and fail the site build
Lets use that to validate our links in our pages for this.

glennawatson replied:
Also I think Microsoft was attempting to achieve something like this with those namespace links https://docs.microsoft.com/en-us/dotnet/api/microsoft.csharp.runtimebinder?view=netcore-2.2

RLittlesII replied:
I can go through the pages again and try and make sure all the links resolve to docs.  Some of that is just how the pages were delivered from Microsoft.

@glennawatson You bring up a good point.  Should we be failing builds for dead hyper links?  I believe it will help us keep the overall health of the site better.

RLittlesII replied:
@glennawatson interestingly enough 
when I add `Settings[DocsKeys.ValidateLinksAsError] = "true";` the build passes
when I add `Settings[DocsKeys.ValidateRelativeLinks] = "true";` the build fails.

glennawatson replied:
DocsKeys.ValidateAbsoluteLinks: System.Boolean

Set to true (the default value is false) to validate all absolute links. Note that this may add considerable time to your generation process.

DocsKeys.ValidateLinksAsError: System.Boolean

Set to true (the default value is false) to report errors on link validation failures.

DocsKeys.ValidateRelativeLinks: System.Boolean

Set to true (the default value) to validate all relative links.

glennawatson replied:
Turn those three to true I guess. If it doesn't slow down the build dramatically let's turn it on in a separate PR today
